### DISCUSSION TITLE
Is there a way to run some code in a package.py at resolve time to decide whether the package should be considered or not?

### DISCUSSION THREAD

**Discussion by herronelou** [AcademySoftwareFoundation/rez#1403] [Q&A]
Hello.

For a lot of our DCCs, we made empty packages that are only pointing at local install files of the software.
Currently, I have a check in the commands section to display an error when the required package is not actually installed locally, but the package still resolves.
I would like to run a little check at resolve time so that if the local install doesn't exist, rez considers the package to not exist, and continues to resolve until it finds a version that satisfies all conditions.

Is there such an option? I can't seem to figure it out from reading the wiki.

**Comment by bryab** [comment]
@herronelou Did you find a solution to this?

I figured out a workaround that kindof works, though it is very hacky.  In an "early" function try to detect if the DCC is installed.  If it isn't, change the variant list such that this package only has a variant for some non-existent variant of another package.  In this case I've chosen to use "arch" because it's a package that is always present.  And I've arbitrarily chosen "disabled" as my non-existent variant.

For example:

```python
@early
def variants():
    if not some_condition:
        return [['arch-disabled']]
    return []

```

A more complete example:

```python

@early
def find_maya():
    # Try to find Maya here and return the path.
    return ""

@early
def variants():
    if not this.find_maya:
        return [['arch-disabled']]
    return []

def commands():
    env.PATH.append(this.find_maya)
```

I'm curious if someone has a better way to handle this.

**Reply by herronelou** [reply]
Sorry for the late reply, 2 years later no, I did not find a better solution. Would love a new feature to do that though.

**Comment by doruchan** [comment]
Hey there,

I wanted to share a workaround that can help with this usecase: I made a `local_{dcc]_install` package that can be required alongside the `dcc` package to say that you want whichever latest dcc version is installed on your machine. Only if no installation is found at all, then it would raise an exception.

It's not meant for production usage, but for building the dcc-related internal packages. It can be used alongside any version constrain for that dcc.

Here is a real life example of the `local_houdini_install` package.py that can be used with the build requires:

```python

name = 'local_houdini_install'

version = "1.0.0"
description = 'Checks which houdini versions are installed on this machine and dynamically require them.'


@early()
def requires():
    """Dynamically require all the versions of Houdini installed.
    """
    req = "|".join(_installed_houdini_versions())
    return [
        f"houdini-{req}",
    ]


@early()
def _installed_houdini_versions():
    """Search for all the Houdini installed on this machine.
    Raise ResourceNotFoundError if no Houdini is installed.
    """
    import glob
    import re
    import sys
    
    version_regex = re.compile(r"\d+.\d+.\d+$")

    if "win" in sys.platform:
        houdini_root_path = "C:/Program Files/Side Effects Software/Houdini "
    else:
        houdini_root_path = "/opt/houdini/hfs"

    installed_versions = []
    for hou_dir in glob.iglob(f"{houdini_root_path}*"):
        match = version_regex.search(hou_dir)
        if match:
            installed_versions.append(match.group(0))

    if not installed_versions:
        from rez.exceptions import ResourceNotFoundError
        raise ResourceNotFoundError("Houdini is not installed on this machine.")
    
    return sorted(installed_versions)


def commands():
    pass


# Nothing to build, just the package.py
build_command = False
```

This example can be used with a generic dcc requirement:
-  `build_requires  = ['houdini', 'local_houdini_install']` 

or with the dcc version ranges specified:
- `build_requires = ['houdini-19-20.5', 'local_houdini_install']`


All credits for the idea to Alex Telford who shared this concept in the rez slack channel.

cheers
